### PR TITLE
Add type for sampled function ids

### PR DESCRIPTION
**Pull request by akopich**
This introduces a type SFID, that represents a sampled function
id and should be used instead `uint64_t` to improve type
readability and catch the bugs at compile time.

This also fixes a bug in the Unit test that is not possible with
SFIDs.

Bug: http://b/235321420
Test: Unit, Compile

### CODE REVIEW

File: src/MizarData/BaselineAndComparisonHelper.cpp [hunk 1]
```diff
@@ -8,53 +8,57 @@
...
+#include <string>
+
 namespace orbit_mizar_data {
 
 template <typename K, typename V>
 [[nodiscard]] static absl::flat_hash_set<V> ValueSet(const absl::flat_hash_map<K, V>& map) {
   absl::flat_hash_set<V> result;
   std::transform(std::begin(map), std::end(map), std::inserter(result, std::begin(result)),
-                 [](const std::pair<K, V> pair) { return pair.second; });
+                 std::mem_fn(&std::pair<std::add_const_t<K>, V>::second));
```
Comment (antonrohr) [inline src/MizarData/BaselineAndComparisonHelper.cpp:19]: (pers pref) I think the original version was easier to read/understand.. what are the benefits of the new version?

File: src/MizarData/BaselineAndComparisonHelper.cpp [hunk 2]
```diff
@@ -8,53 +8,57 @@
...
+  absl::flat_hash_set<std::string> comparison_names =
+      ValueSet<uint64_t, std::string>(comparison_address_to_name);
 
-  absl::flat_hash_map<std::string, uint64_t> name_to_frame_id;
-  absl::flat_hash_map<uint64_t, std::string> frame_id_to_name;
+  absl::flat_hash_map<std::string, SFID> name_to_sfid;
+  absl::flat_hash_map<SFID, std::string> sfid_to_name;
 
-  uint64_t next_frame_id = 1;
+  uint64_t next_sfid_value = 1;
```
Comment (antonrohr) [inline src/MizarData/BaselineAndComparisonHelper.cpp:46]: you could consider implementing operator++ for SFID and use it here instead of uint64_t

**Review by antonrohr — APPROVED** [review]

File: src/MizarData/BaselineAndComparisonHelper.cpp (hunk 2 above)
Comment (akopich) [reply src/MizarData/BaselineAndComparisonHelper.cpp:46]: True. But I thought if SFID as of smth immutable, that is not necessarily consecutive and neither backed with uint64_t. Let's see what other reviewers think.

File: src/MizarData/BaselineAndComparisonHelper.cpp (hunk 1 above)
Comment (akopich) [reply src/MizarData/BaselineAndComparisonHelper.cpp:19]: > what are the benefits of the new version?
I had fun writing it. 

I guess, you're right and it should be lambda here.

File: src/MizarData/MizarPairedDataTest.cpp [hunk 3]
```diff
@@ -78,22 +78,21 @@ const std::unique_ptr<orbit_client_data::CallstackData> kCallstackData = [] {
...
-    const std::vector<uint64_t>& addresses) {
+[[nodiscard]] static std::vector<SFID> SFIDsForCallstacks(const std::vector<uint64_t>& addresses) {
   std::vector<uint64_t> good_addresses;
   std::copy_if(std::begin(addresses), std::end(addresses), std::back_inserter(good_addresses),
                [](uint64_t address) { return kAddressToId.contains(address); });
-  std::vector<uint64_t> ids;
+  std::vector<SFID> ids;
   std::transform(std::begin(good_addresses), std::end(good_addresses), std::back_inserter(ids),
-                 [](uint64_t address) { return kAddressToId.at(address); });
+                 [](uint64_t address) { return SFID(kAddressToId.at(address)); });
```
[the remark this answers is not in the document]
Comment (akopich) [reply src/MizarData/MizarPairedDataTest.cpp:87]: Thank you, you are right. But why is that so? I've marked the SFID constructor as explicit.

[the remark this answers is not in the document]
Comment (akopich) [reply src/MizarData/MizarPairedDataTest.cpp:87]: ... and my goal was to make implicit conversions between SFID and int types impossible.

File: src/MizarData/BaselineAndComparisonHelper.cpp (hunk 2 above)
Comment (florian-kuebler) [reply src/MizarData/BaselineAndComparisonHelper.cpp:46]: I am not completely sure about this (and probably also not sure about the whole change).
For the concrete situation, I see the argument that SFID is something immutable and does not necessarily have all the properties of natural numbers. Also, if we start adding a ++ function, we quickly also want other numerical operations (+,-,*,/).
If we end up like this, we will have a quite large wrapper to maintain around uint64 just for the purpose of better type safty (which I am in general a big friend of).

I mean what we would really like to have here is a strong typedef, right? I.e. I can only pass in uint64's that I explicitly declared as SFID. Unfortunately this is not really possible.

I'd like to have @beckerhe's opinion on the whole change.

Comment (florian-kuebler) [reply src/MizarData/BaselineAndComparisonHelper.cpp:46]: Yeah, I am not really sure if this is all worth it :/

Comment (akopich) [reply src/MizarData/BaselineAndComparisonHelper.cpp:46]: At least, now we have `flat_hash_map<uint64_t, SFID>`, `flat_hash_map<uint64_t, std::string>` and `flat_hash_map<SFID, std::string>`.
Not `flat_hash_map<uint64_t, uint64_t>`, `flat_hash_map<uint64_t, std::string>` and `flat_hash_map<uint64_t, std::string>`

File: src/MizarData/include/MizarData/SampledFunctionId.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace orbit_mizar_data {
+
+// The class represents a sampled function id. These ids are the same for the same function across
+// all the captures.
+class SFID {
+ public:
+  constexpr explicit SFID(uint64_t id) : id_(id) {}
+
+  friend bool operator==(const SFID& lhs, const SFID& rhs) { return lhs.id_ == rhs.id_; }
```
Comment (beckerhe) [inline src/MizarData/include/MizarData/SampledFunctionId.h:22]: nodiscard

Comment (beckerhe) [reply src/MizarData/include/MizarData/SampledFunctionId.h:22]: For symmetry reasons you should also define `operator!=`

File: src/MizarData/BaselineAndComparisonHelper.cpp (hunk 2 above)
Comment (florian-kuebler) [reply src/MizarData/BaselineAndComparisonHelper.cpp:46]: But this could be also archived with a simple typedef, right?

Comment (akopich) [reply src/MizarData/BaselineAndComparisonHelper.cpp:46]: True. OTOH, now we have surprising conversions only in some places. With typedefs, these are going to be the same type.

Comment (beckerhe) [reply src/MizarData/BaselineAndComparisonHelper.cpp:46]: > I am not completely sure about this (and probably also not sure about the whole change). For the concrete situation, I see the argument that SFID is something immutable and does not necessarily have all the properties of natural numbers. Also, if we start adding a ++ function, we quickly also want other numerical operations (+,-,*,/). If we end up like this, we will have a quite large wrapper to maintain around uint64 just for the purpose of better type safty (which I am in general a big friend of).
> 
> I mean what we would really like to have here is a strong typedef, right? I.e. I can only pass in uint64's that I explicitly declared as SFID. Unfortunately this is not really possible.
> 
> I'd like to have @beckerhe's opinion on the whole change.

I see it the following way:

- It's great to have SFID as a separate type to make clear it's something else than a natural number
- It would be totally possible to provide operator++ for this type, but then we should make the type properly arithmetic and provide other operations as well.
- But I don't think that's what we want because SFID is still an ID and not a number. It just happens to be based on an integer, but technically it could also use a string as the underlying type.
- That being said, to create new SFIDs it's necessary to understand the underlying type. That's why I like to keep this part as an uint64_t. SFID itself has no notion of a next ID because it doesn't offer a measure of distance or greater smaller comparison. All these things need to be done in the underlying type - which forces the user to understand what they need to do to create a new id.

TLDR: I think I like it that way. It makes sense from a code ownership and responsibility sharing perspective. Personally I would have provided a getter for the value in SFID and then kept the last used SFID around and created a new one like this `SFID{previous_sfid.value() + 1}` but that's really just preference.

**Review by beckerhe — COMMENTED** [review]
Not yet fully through the change, but wanted to get out the comments before the next meeting

File: src/MizarData/MizarPairedDataTest.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (antonrohr) [reply src/MizarData/MizarPairedDataTest.cpp:87]: I think you accomplished the goal. This is not doing any conversions. `kAddressToId` is already a map from `uint64_t` to `SFID`

File: src/MizarData/BaselineAndComparisonHelper.cpp (hunk 1 above)
Comment (antonrohr) [reply src/MizarData/BaselineAndComparisonHelper.cpp:19]: fwiw, I had fun trying to understand it

File: src/MizarData/BaselineAndComparisonHelper.cpp (hunk 2 above)
Comment (antonrohr) [reply src/MizarData/BaselineAndComparisonHelper.cpp:46]: > So we can't really prevent that conversion.

Wait. I think you accomplished that. A typedef would have the downside of implicit conversions, you approach with the explicit constructor does not. 

My initial comment was really meant as as a suggestion to think about and not as a request to change. I like the argumentation around SFID not being an arithmetic type and I would leave everything as is, or do the suggestion that hebecker had (provide .value()). If you do that, you could also consider removing the `operator std::string`, which is currently the only way of getting the `id_` out of `SFID`.

Comment (akopich) [reply src/MizarData/BaselineAndComparisonHelper.cpp:46]: @antonrohr , thank you, my bad. Conversion is prevented, everything's fine. 

I don't really like `.value()` method, as its return type would force us to keep the think uint64_t-based.

**Review by florian-kuebler — APPROVED** [review]
LGTM, thanks!

**Review by beckerhe — APPROVED** [review]
